# Titanic competition with TensorFlow Decision Forests

This notebook will take you through the steps needed to train a baseline Gradient Boosted Trees Model using TensorFlow Decision Forests and creating a submission on the Titanic competition. 

This notebook shows:

1. How to do some basic pre-processing. For example, the passenger names will be tokenized, and ticket names will be splitted in parts.
1. How to train a Gradient Boosted Trees (GBT) with default parameters
1. How to train a GBT with improved default parameters
1. How to tune the parameters of a GBTs
1. How to train and ensemble many GBTs

# Imports dependencies

In [1]:
import numpy as np
import pandas as pd
import os

import tensorflow as tf
import tensorflow_decision_forests as tfdf

print(f"Found TF-DF {tfdf.__version__}")

Found TF-DF 1.2.0


# Load dataset

In [2]:
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
serving_df = pd.read_csv("/kaggle/input/titanic/test.csv")

train_df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

# Prepare dataset

We will apply the following transformations on the dataset.

1. Tokenize the names. For example, "Braund, Mr. Owen Harris" will become ["Braund", "Mr.", "Owen", "Harris"].
2. Extract any prefix in the ticket. For example ticket "STON/O2. 3101282" will become "STON/O2." and 3101282.

In [3]:
def preprocess(df):
    df = df.copy()
    
    def normalize_name(x):
        return " ".join([v.strip(",()[].\"'") for v in x.split(" ")])
    
    def ticket_number(x):
        return x.split(" ")[-1]
        
    def ticket_item(x):
        items = x.split(" ")
        if len(items) == 1:
            return "NONE"
        return "_".join(items[0:-1])
    
    df["Name"] = df["Name"].apply(normalize_name)
    df["Ticket_number"] = df["Ticket"].apply(ticket_number)
    df["Ticket_item"] = df["Ticket"].apply(ticket_item)                     
    return df
    
preprocessed_train_df = preprocess(train_df)
preprocessed_serving_df = preprocess(serving_df)

preprocessed_train_df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                              Name     Sex   Age  SibSp  \
0                            Braund Mr Owen Harris    male  22.0      1   
1  Cumings Mrs John Bradley Florence Briggs Thayer  female  38.0      1   
2                             Heikkinen Miss Laina  female  26.0      0   
3         Futrelle Mrs Jacques Heath Lily May Peel  female  35.0      1   
4                           Allen Mr William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked Ticket_number Ticket_item  
0      0         A/5 21171   7.2500   NaN        S         21171         A/5  
1      0          PC 17599  71.2833   C85        C         17599          PC  
2      0  STON/O2. 3101282   7.9250   NaN        S       3101282    STON/O2.  
3      0            113803  53.1000  C123        S        113803        NONE  
4      0            373450   8.0500   NaN        S        373450        NONE

Let's keep the list of the input features of the model. Notably, we don't want to train our model on the "PassengerId" and "Ticket" features.

In [4]:
input_features = list(preprocessed_train_df.columns)
input_features.remove("Ticket")
input_features.remove("PassengerId")
input_features.remove("Survived")
#input_features.remove("Ticket_number")

print(f"Input features: {input_features}")

Input features: ['Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin', 'Embarked', 'Ticket_number', 'Ticket_item']


# Convert Pandas dataset to TensorFlow Dataset

In [5]:
def tokenize_names(features, labels=None):
    """Divite the names into tokens. TF-DF can consume text tokens natively."""
    features["Name"] =  tf.strings.split(features["Name"])
    return features, labels

train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_train_df,label="Survived").map(tokenize_names)
serving_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_serving_df).map(tokenize_names)

# Train model with default parameters

### Train model

First, we are training a GradientBoostedTreesModel model with the default parameters.

In [6]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    random_seed=1234,
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2023-08-30T07:16:19.248207395+00:00 kernel.cc:1214] Loading model from path /tmp/tmpn1mth73r/model/ with prefix 0820169111824695
[INFO 2023-08-30T07:16:19.257904222+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-30T07:16:19.258048745+00:00 kernel.cc:1046] Use fast generic engine


Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: could not get source code
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Accuracy: 0.8260869383811951 Loss:0.8608942627906799


# Train model with improved default parameters

Now you'll use some specific parameters when creating the GBT model

In [7]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    
    #num_trees=2000,
    
    # Only for GBT.
    # A bit slower, but great to understand the model.
    # compute_permutation_variable_importance=True,
    
    # Change the default hyper-parameters
    # hyperparameter_template="benchmark_rank1@v1",
    
    #num_trees=1000,
    #tuner=tuner
    
    min_examples=1,
    categorical_algorithm="RANDOM",
    #max_depth=4,
    shrinkage=0.05,
    #num_candidate_attributes_ratio=0.2,
    split_axis="SPARSE_OBLIQUE",
    sparse_oblique_normalization="MIN_MAX",
    sparse_oblique_num_projections_exponent=2.0,
    num_trees=2000,
    #validation_ratio=0.0,
    random_seed=1234,
    
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2023-08-30T07:16:23.061511004+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbrtkybsr/model/ with prefix 4750e3fd95e14722
[INFO 2023-08-30T07:16:23.070406742+00:00 decision_forest.cc:661] Model loaded with 33 root(s), 1823 node(s), and 10 input feature(s).
[INFO 2023-08-30T07:16:23.070603051+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.760869562625885 Loss:1.0154211521148682


Let's look at the model and you can also notice the information about variable importance that the model figured out

In [8]:
model.summary()

Model: "gradient_boosted_trees_model_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
Total params: 1
Trainable params: 0
Non-trainable params: 1
_________________________________________________________________
Type: "GRADIENT_BOOSTED_TREES"
Task: CLASSIFICATION
Label: "__LABEL"

Input Features (11):
	Age
	Cabin
	Embarked
	Fare
	Name
	Parch
	Pclass
	Sex
	SibSp
	Ticket_item
	Ticket_number

No weights

Variable Importance: INV_MEAN_MIN_DEPTH:
    1.           "Sex"  0.576632 ################
    2.           "Age"  0.364297 #######
    3.          "Fare"  0.278839 ####
    4.          "Name"  0.208548 #
    5. "Ticket_number"  0.180792 
    6.        "Pclass"  0.176962 
    7.         "Parch"  0.176659 
    8.   "Ticket_item"  0.175540 
    9.      "Embarked"  0.172339 
   10.         "SibSp"  0.170442 

Variable Importance: NUM_AS_ROOT:
    1.  "Sex" 28.000000 ################
    2. "Name"  5.000000 

# Make predictions

In [9]:
def prediction_to_kaggle_format(model, threshold=0.5):
    proba_survive = model.predict(serving_ds, verbose=0)[:,0]
    return pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (proba_survive >= threshold).astype(int)
    })

def make_submission(kaggle_predictions):
    path="/kaggle/working/submission.csv"
    kaggle_predictions.to_csv(path, index=False)
    print(f"Submission exported to {path}")
    
kaggle_predictions = prediction_to_kaggle_format(model)
make_submission(kaggle_predictions)
!head /kaggle/working/submission.csv

Submission exported to /kaggle/working/submission.csv
PassengerId,Survived
892,0
893,0
894,0
895,0
896,0
897,0
898,0
899,0
900,1


# Training a model with hyperparameter tunning

Hyper-parameter tuning is enabled by specifying the tuner constructor argument of the model. The tuner object contains all the configuration of the tuner (search space, optimizer, trial and objective).


In [10]:
tuner = tfdf.tuner.RandomSearch(num_trials=1000)
tuner.choice("min_examples", [2, 5, 7, 10])
tuner.choice("categorical_algorithm", ["CART", "RANDOM"])

local_search_space = tuner.choice("growing_strategy", ["LOCAL"])
local_search_space.choice("max_depth", [3, 4, 5, 6, 8])

global_search_space = tuner.choice("growing_strategy", ["BEST_FIRST_GLOBAL"], merge=True)
global_search_space.choice("max_num_nodes", [16, 32, 64, 128, 256])

#tuner.choice("use_hessian_gain", [True, False])
tuner.choice("shrinkage", [0.02, 0.05, 0.10, 0.15])
tuner.choice("num_candidate_attributes_ratio", [0.2, 0.5, 0.9, 1.0])


tuner.choice("split_axis", ["AXIS_ALIGNED"])
oblique_space = tuner.choice("split_axis", ["SPARSE_OBLIQUE"], merge=True)
oblique_space.choice("sparse_oblique_normalization",
                     ["NONE", "STANDARD_DEVIATION", "MIN_MAX"])
oblique_space.choice("sparse_oblique_weights", ["BINARY", "CONTINUOUS"])
oblique_space.choice("sparse_oblique_num_projections_exponent", [1.0, 1.5])

# Tune the model. Notice the `tuner=tuner`.
tuned_model = tfdf.keras.GradientBoostedTreesModel(tuner=tuner)
tuned_model.fit(train_ds, verbose=0)

tuned_self_evaluation = tuned_model.make_inspector().evaluation()
print(f"Accuracy: {tuned_self_evaluation.accuracy} Loss:{tuned_self_evaluation.loss}")

Use /tmp/tmpqa_cjt82 as temporary training directory


[INFO 2023-08-30T07:18:26.001473501+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqa_cjt82/model/ with prefix 1941c9125abe4390
[INFO 2023-08-30T07:18:26.015713171+00:00 decision_forest.cc:661] Model loaded with 19 root(s), 589 node(s), and 12 input feature(s).
[INFO 2023-08-30T07:18:26.015768396+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesGeneric" built
[INFO 2023-08-30T07:18:26.015799216+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.9178082346916199 Loss:0.6503586769104004


In the last line in the cell above, you can see the accuracy is higher than previously with default parameters and parameters set by hand.

This is the main idea behing hyperparameter tuning.

For more information you can follow this tutorial: [Automated hyper-parameter tuning](https://www.tensorflow.org/decision_forests/tutorials/automatic_tuning_colab)

# Making an ensemble

Here you'll create 100 models with different seeds and combine their results

This approach removes a little bit the random aspects related to creating ML models

In the GBT creation is used the `honest` parameter. It will use different training examples to infer the structure and the leaf values. This regularization technique trades examples for bias estimates.

In [11]:
predictions = None
num_predictions = 0

for i in range(100):
    print(f"i:{i}")
    # Possible models: GradientBoostedTreesModel or RandomForestModel
    model = tfdf.keras.GradientBoostedTreesModel(
        verbose=0, # Very few logs
        features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
        exclude_non_specified_features=True, # Only use the features in "features"

        #min_examples=1,
        #categorical_algorithm="RANDOM",
        ##max_depth=4,
        #shrinkage=0.05,
        ##num_candidate_attributes_ratio=0.2,
        #split_axis="SPARSE_OBLIQUE",
        #sparse_oblique_normalization="MIN_MAX",
        #sparse_oblique_num_projections_exponent=2.0,
        #num_trees=2000,
        ##validation_ratio=0.0,
        random_seed=i,
        honest=True,
    )
    model.fit(train_ds)
    
    sub_predictions = model.predict(serving_ds, verbose=0)[:,0]
    if predictions is None:
        predictions = sub_predictions
    else:
        predictions += sub_predictions
    num_predictions += 1

predictions/=num_predictions

kaggle_predictions = pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (predictions >= 0.5).astype(int)
    })

make_submission(kaggle_predictions)

i:0


[INFO 2023-08-30T07:18:26.944448925+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6k33ajhe/model/ with prefix eabf9ad5b1814240
[INFO 2023-08-30T07:18:26.948242858+00:00 kernel.cc:1046] Use fast generic engine


i:1


[INFO 2023-08-30T07:18:28.33393016+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2a9djc7e/model/ with prefix 49342e28e9d44e30
[INFO 2023-08-30T07:18:28.350972764+00:00 kernel.cc:1046] Use fast generic engine


i:2


[INFO 2023-08-30T07:18:29.285887944+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgz7o390t/model/ with prefix 389c984394c34cff
[INFO 2023-08-30T07:18:29.290049067+00:00 kernel.cc:1046] Use fast generic engine


i:3


[INFO 2023-08-30T07:18:30.738477802+00:00 kernel.cc:1214] Loading model from path /tmp/tmpry0gwkic/model/ with prefix ad9f72a4396a415c
[INFO 2023-08-30T07:18:30.76189722+00:00 kernel.cc:1046] Use fast generic engine


i:4


[INFO 2023-08-30T07:18:31.762484402+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6mrxu0ht/model/ with prefix ec641a8007654cc4
[INFO 2023-08-30T07:18:31.767554283+00:00 kernel.cc:1046] Use fast generic engine


i:5


[INFO 2023-08-30T07:18:32.703591423+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7_3yofj8/model/ with prefix c29cf45a67cd496e
[INFO 2023-08-30T07:18:32.706271829+00:00 kernel.cc:1046] Use fast generic engine


i:6


[INFO 2023-08-30T07:18:33.703732319+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2os_sads/model/ with prefix df5c169a44c04a6a
[INFO 2023-08-30T07:18:33.710731936+00:00 kernel.cc:1046] Use fast generic engine


i:7


[INFO 2023-08-30T07:18:35.007601031+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnrrn__vz/model/ with prefix 9728894dd387403d
[INFO 2023-08-30T07:18:35.025401044+00:00 kernel.cc:1046] Use fast generic engine


i:8


[INFO 2023-08-30T07:18:36.438108329+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_eora67a/model/ with prefix cc16d6796c4a445b
[INFO 2023-08-30T07:18:36.44731138+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-30T07:18:36.447356771+00:00 kernel.cc:1046] Use fast generic engine


i:9


[INFO 2023-08-30T07:18:37.601068503+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgf6_kc2n/model/ with prefix 2b86b73aafc443a6
[INFO 2023-08-30T07:18:37.614667363+00:00 kernel.cc:1046] Use fast generic engine


i:10


[INFO 2023-08-30T07:18:38.571925626+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5jabx5of/model/ with prefix bff7d9ede94346c9
[INFO 2023-08-30T07:18:38.577179584+00:00 kernel.cc:1046] Use fast generic engine


i:11


[INFO 2023-08-30T07:18:39.773810933+00:00 kernel.cc:1214] Loading model from path /tmp/tmp250pf4vg/model/ with prefix 1799499cd15e41bd
[INFO 2023-08-30T07:18:39.786882296+00:00 kernel.cc:1046] Use fast generic engine


i:12


[INFO 2023-08-30T07:18:40.733865105+00:00 kernel.cc:1214] Loading model from path /tmp/tmps0bq7xoy/model/ with prefix 2a2fc0bc568e4170
[INFO 2023-08-30T07:18:40.739145545+00:00 kernel.cc:1046] Use fast generic engine


i:13


[INFO 2023-08-30T07:18:41.892158617+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5o_oyut6/model/ with prefix 1ec82c6920884e1e
[INFO 2023-08-30T07:18:41.902438125+00:00 kernel.cc:1046] Use fast generic engine


i:14


[INFO 2023-08-30T07:18:42.943556666+00:00 kernel.cc:1214] Loading model from path /tmp/tmpy2kg0w70/model/ with prefix 35a6675ab8a2439c
[INFO 2023-08-30T07:18:42.949327438+00:00 kernel.cc:1046] Use fast generic engine


i:15


[INFO 2023-08-30T07:18:43.946750209+00:00 kernel.cc:1214] Loading model from path /tmp/tmpf41ekowp/model/ with prefix 24001303974d4d69
[INFO 2023-08-30T07:18:43.953671303+00:00 kernel.cc:1046] Use fast generic engine


i:16


[INFO 2023-08-30T07:18:45.078817375+00:00 kernel.cc:1214] Loading model from path /tmp/tmpka_kfb_2/model/ with prefix 29c2f73054ec4467
[INFO 2023-08-30T07:18:45.090341316+00:00 kernel.cc:1046] Use fast generic engine


i:17


[INFO 2023-08-30T07:18:46.238149073+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxa5eamc3/model/ with prefix 6d7801d159e34d17
[INFO 2023-08-30T07:18:46.250491625+00:00 kernel.cc:1046] Use fast generic engine


i:18


[INFO 2023-08-30T07:18:47.351945971+00:00 kernel.cc:1214] Loading model from path /tmp/tmpg1wa14y9/model/ with prefix 21ac972bf0ba4900
[INFO 2023-08-30T07:18:47.362557119+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-30T07:18:47.362596164+00:00 kernel.cc:1046] Use fast generic engine


i:19


[INFO 2023-08-30T07:18:48.623159255+00:00 kernel.cc:1214] Loading model from path /tmp/tmpk2r0a7ap/model/ with prefix 57e661bad31941fb
[INFO 2023-08-30T07:18:48.639347111+00:00 kernel.cc:1046] Use fast generic engine


i:20


[INFO 2023-08-30T07:18:50.24272411+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqkdak7oy/model/ with prefix 11a0e252752d4547
[INFO 2023-08-30T07:18:50.25616659+00:00 kernel.cc:1046] Use fast generic engine


i:21


[INFO 2023-08-30T07:18:51.203817665+00:00 kernel.cc:1214] Loading model from path /tmp/tmpyhzqypgr/model/ with prefix 753cfbf7de794282
[INFO 2023-08-30T07:18:51.208568293+00:00 kernel.cc:1046] Use fast generic engine


i:22


[INFO 2023-08-30T07:18:52.172720794+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvecg5f8t/model/ with prefix 79b5b68a676c4cee
[INFO 2023-08-30T07:18:52.177872319+00:00 kernel.cc:1046] Use fast generic engine


i:23


[INFO 2023-08-30T07:18:53.315155999+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7korku7o/model/ with prefix 7de36570c1244abe
[INFO 2023-08-30T07:18:53.323120927+00:00 kernel.cc:1046] Use fast generic engine


i:24


[INFO 2023-08-30T07:18:54.290021696+00:00 kernel.cc:1214] Loading model from path /tmp/tmpb8ild90s/model/ with prefix 20b5058dc11147ac
[INFO 2023-08-30T07:18:54.295292955+00:00 kernel.cc:1046] Use fast generic engine


i:25


[INFO 2023-08-30T07:18:55.395533775+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2dkddxud/model/ with prefix e62965166e1245d5
[INFO 2023-08-30T07:18:55.405744456+00:00 kernel.cc:1046] Use fast generic engine


i:26


[INFO 2023-08-30T07:18:56.464155154+00:00 kernel.cc:1214] Loading model from path /tmp/tmpg46x9xbw/model/ with prefix 4fbf87f7c0144e4f
[INFO 2023-08-30T07:18:56.473422089+00:00 kernel.cc:1046] Use fast generic engine


i:27


[INFO 2023-08-30T07:18:58.06096153+00:00 kernel.cc:1214] Loading model from path /tmp/tmplv49_xil/model/ with prefix 6f68f3be67a44a16
[INFO 2023-08-30T07:18:58.066971228+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-30T07:18:58.067016588+00:00 kernel.cc:1046] Use fast generic engine


i:28


[INFO 2023-08-30T07:18:59.073373535+00:00 kernel.cc:1214] Loading model from path /tmp/tmpd9klmke9/model/ with prefix 2bfbef8d40b64782
[INFO 2023-08-30T07:18:59.07823121+00:00 kernel.cc:1046] Use fast generic engine


i:29


[INFO 2023-08-30T07:19:00.254735486+00:00 kernel.cc:1214] Loading model from path /tmp/tmppw9aiw6z/model/ with prefix 590871080a5047f6
[INFO 2023-08-30T07:19:00.266550323+00:00 kernel.cc:1046] Use fast generic engine


i:30


[INFO 2023-08-30T07:19:02.263988308+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgzx409ru/model/ with prefix 750830b352c0468a
[INFO 2023-08-30T07:19:02.2933545+00:00 kernel.cc:1046] Use fast generic engine


i:31


[INFO 2023-08-30T07:19:03.371739619+00:00 kernel.cc:1214] Loading model from path /tmp/tmpv1jxtfuz/model/ with prefix 7b06e8fb1ac841e4
[INFO 2023-08-30T07:19:03.381124092+00:00 kernel.cc:1046] Use fast generic engine


i:32


[INFO 2023-08-30T07:19:04.343444427+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsm2ttx7u/model/ with prefix 7d533c78fef541e6
[INFO 2023-08-30T07:19:04.348760793+00:00 kernel.cc:1046] Use fast generic engine


i:33


[INFO 2023-08-30T07:19:05.492691583+00:00 kernel.cc:1214] Loading model from path /tmp/tmphfk12yb9/model/ with prefix 8123f686a6584da5
[INFO 2023-08-30T07:19:05.504819067+00:00 kernel.cc:1046] Use fast generic engine


i:34


[INFO 2023-08-30T07:19:06.527988983+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcs8exeyx/model/ with prefix 4f9015664d6b42e1
[INFO 2023-08-30T07:19:06.53589297+00:00 kernel.cc:1046] Use fast generic engine


i:35


[INFO 2023-08-30T07:19:07.540807661+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsw4oac4p/model/ with prefix ed06570a6aa640fa
[INFO 2023-08-30T07:19:07.548190976+00:00 kernel.cc:1046] Use fast generic engine


i:36


[INFO 2023-08-30T07:19:08.73009445+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxzxt1jpt/model/ with prefix e1554a7a43b64b9d
[INFO 2023-08-30T07:19:08.742883863+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-30T07:19:08.742925429+00:00 kernel.cc:1046] Use fast generic engine


i:37


[INFO 2023-08-30T07:19:09.808739973+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkp3f0v_q/model/ with prefix 63f4792c01b34137
[INFO 2023-08-30T07:19:09.817049281+00:00 kernel.cc:1046] Use fast generic engine


i:38


[INFO 2023-08-30T07:19:10.959749357+00:00 kernel.cc:1214] Loading model from path /tmp/tmpit5ft3fj/model/ with prefix deb6687c412c4fa6
[INFO 2023-08-30T07:19:10.971733042+00:00 kernel.cc:1046] Use fast generic engine


i:39


[INFO 2023-08-30T07:19:12.112433898+00:00 kernel.cc:1214] Loading model from path /tmp/tmpc6ru4c77/model/ with prefix 720b2dd6d0b9433c
[INFO 2023-08-30T07:19:12.124068489+00:00 kernel.cc:1046] Use fast generic engine


i:40


[INFO 2023-08-30T07:19:13.659668076+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnpd9eofr/model/ with prefix 2322a464be6c4a08
[INFO 2023-08-30T07:19:13.664067067+00:00 kernel.cc:1046] Use fast generic engine


i:41


[INFO 2023-08-30T07:19:14.896809972+00:00 kernel.cc:1214] Loading model from path /tmp/tmplvrle5w8/model/ with prefix 3096ded72edd4f85
[INFO 2023-08-30T07:19:14.910759385+00:00 kernel.cc:1046] Use fast generic engine


i:42


[INFO 2023-08-30T07:19:16.451589576+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjxvopirq/model/ with prefix 783f5c78f7d44020
[INFO 2023-08-30T07:19:16.460028717+00:00 kernel.cc:1046] Use fast generic engine


i:43


[INFO 2023-08-30T07:19:17.725555112+00:00 kernel.cc:1214] Loading model from path /tmp/tmprfmqo1k8/model/ with prefix b98ca17f2f5e427b
[INFO 2023-08-30T07:19:17.741228078+00:00 kernel.cc:1046] Use fast generic engine


i:44


[INFO 2023-08-30T07:19:18.786508875+00:00 kernel.cc:1214] Loading model from path /tmp/tmpazanr5jh/model/ with prefix eea64c507cec4547
[INFO 2023-08-30T07:19:18.795159753+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-30T07:19:18.795206511+00:00 kernel.cc:1046] Use fast generic engine


i:45


[INFO 2023-08-30T07:19:19.734166251+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfg6rgm0h/model/ with prefix 4d59efe704854068
[INFO 2023-08-30T07:19:19.73747757+00:00 kernel.cc:1046] Use fast generic engine


i:46


[INFO 2023-08-30T07:19:20.924234016+00:00 kernel.cc:1214] Loading model from path /tmp/tmpej0x8chr/model/ with prefix 363fb30888054c14
[INFO 2023-08-30T07:19:20.93733607+00:00 kernel.cc:1046] Use fast generic engine


i:47


[INFO 2023-08-30T07:19:22.087697218+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgxmed4j6/model/ with prefix 2bda2996963f462a
[INFO 2023-08-30T07:19:22.099637221+00:00 kernel.cc:1046] Use fast generic engine


i:48


[INFO 2023-08-30T07:19:23.558762839+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5q128wwh/model/ with prefix c6ba28984dd7420d
[INFO 2023-08-30T07:19:23.562859753+00:00 kernel.cc:1046] Use fast generic engine


i:49


[INFO 2023-08-30T07:19:24.686647057+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnoiv5fht/model/ with prefix b52328caae9f47c0
[INFO 2023-08-30T07:19:24.692524378+00:00 kernel.cc:1046] Use fast generic engine


i:50


[INFO 2023-08-30T07:19:25.801804682+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxev3bknr/model/ with prefix 50f0ee52fe604e3d
[INFO 2023-08-30T07:19:25.812845153+00:00 kernel.cc:1046] Use fast generic engine


i:51


[INFO 2023-08-30T07:19:27.038571672+00:00 kernel.cc:1214] Loading model from path /tmp/tmpuoa1q3qr/model/ with prefix 2e614b7d1f15401e
[INFO 2023-08-30T07:19:27.052492291+00:00 kernel.cc:1046] Use fast generic engine


i:52


[INFO 2023-08-30T07:19:28.077638553+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnrav1kfd/model/ with prefix e5986fa22dc24754
[INFO 2023-08-30T07:19:28.084805458+00:00 kernel.cc:1046] Use fast generic engine


i:53


[INFO 2023-08-30T07:19:29.120913033+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmxz0kwi6/model/ with prefix d3e39d7c9d104c07
[INFO 2023-08-30T07:19:29.127646191+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-30T07:19:29.127681402+00:00 kernel.cc:1046] Use fast generic engine


i:54


[INFO 2023-08-30T07:19:30.655020555+00:00 kernel.cc:1214] Loading model from path /tmp/tmplhwj8p21/model/ with prefix d5363d54f1634b02
[INFO 2023-08-30T07:19:30.659163495+00:00 kernel.cc:1046] Use fast generic engine


i:55


[INFO 2023-08-30T07:19:31.915096251+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgfwfcvk_/model/ with prefix a884bc1d5feb4998
[INFO 2023-08-30T07:19:31.930713959+00:00 kernel.cc:1046] Use fast generic engine


i:56


[INFO 2023-08-30T07:19:33.553470233+00:00 kernel.cc:1214] Loading model from path /tmp/tmpg52go1cz/model/ with prefix 2dbedd42865840d5
[INFO 2023-08-30T07:19:33.564156713+00:00 kernel.cc:1046] Use fast generic engine


i:57


[INFO 2023-08-30T07:19:35.12283182+00:00 kernel.cc:1214] Loading model from path /tmp/tmpazqoxszy/model/ with prefix 278e7333df2a4cf4
[INFO 2023-08-30T07:19:35.127013627+00:00 kernel.cc:1046] Use fast generic engine


i:58


[INFO 2023-08-30T07:19:36.248047421+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6x1qyci_/model/ with prefix 2842bc69d1eb47e6
[INFO 2023-08-30T07:19:36.25420966+00:00 kernel.cc:1046] Use fast generic engine


i:59


[INFO 2023-08-30T07:19:37.402100316+00:00 kernel.cc:1214] Loading model from path /tmp/tmpx5s_o721/model/ with prefix 5858561566144268
[INFO 2023-08-30T07:19:37.410736481+00:00 kernel.cc:1046] Use fast generic engine


i:60


[INFO 2023-08-30T07:19:39.129622475+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcgvdeulh/model/ with prefix 21ed8c5fdb2146d3
[INFO 2023-08-30T07:19:39.138691258+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-30T07:19:39.138740009+00:00 kernel.cc:1046] Use fast generic engine


i:61


[INFO 2023-08-30T07:19:40.143582927+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnypasrjn/model/ with prefix 4289ea931b6c4708
[INFO 2023-08-30T07:19:40.148426517+00:00 kernel.cc:1046] Use fast generic engine


i:62


[INFO 2023-08-30T07:19:41.641056169+00:00 kernel.cc:1214] Loading model from path /tmp/tmpw9agzjfn/model/ with prefix 6029c08ed2f744bf
[INFO 2023-08-30T07:19:41.664263921+00:00 kernel.cc:1046] Use fast generic engine


i:63


[INFO 2023-08-30T07:19:42.921776122+00:00 kernel.cc:1214] Loading model from path /tmp/tmpk_y0tymi/model/ with prefix 3d8d78e5592d4058
[INFO 2023-08-30T07:19:42.93026445+00:00 kernel.cc:1046] Use fast generic engine


i:64


[INFO 2023-08-30T07:19:43.940672523+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2r1xj6lf/model/ with prefix b0c84d3cd11c4750
[INFO 2023-08-30T07:19:43.947490972+00:00 kernel.cc:1046] Use fast generic engine


i:65


[INFO 2023-08-30T07:19:44.985912111+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfv5nagqo/model/ with prefix 75bf2cc116994633
[INFO 2023-08-30T07:19:44.990942699+00:00 kernel.cc:1046] Use fast generic engine


i:66


[INFO 2023-08-30T07:19:45.992020843+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcqlo00h2/model/ with prefix b821247aaab24e2a
[INFO 2023-08-30T07:19:45.997986633+00:00 kernel.cc:1046] Use fast generic engine


i:67


[INFO 2023-08-30T07:19:47.317528763+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgb68iv49/model/ with prefix 220a7274c7474bb7
[INFO 2023-08-30T07:19:47.334493931+00:00 kernel.cc:1046] Use fast generic engine


i:68


[INFO 2023-08-30T07:19:48.488393593+00:00 kernel.cc:1214] Loading model from path /tmp/tmpoko4fsvm/model/ with prefix 30b33fa5d380489c
[INFO 2023-08-30T07:19:48.499091684+00:00 kernel.cc:1046] Use fast generic engine


i:69


[INFO 2023-08-30T07:19:49.549962633+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq8qppe9w/model/ with prefix 9cffbe8e15c24e2e
[INFO 2023-08-30T07:19:49.556130057+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-30T07:19:49.556170904+00:00 kernel.cc:1046] Use fast generic engine


i:70


[INFO 2023-08-30T07:19:50.629848537+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnoy_nl00/model/ with prefix 49a2a296a20b4d93
[INFO 2023-08-30T07:19:50.637792678+00:00 kernel.cc:1046] Use fast generic engine


i:71


[INFO 2023-08-30T07:19:51.684778453+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqvppzglm/model/ with prefix 5a044070a9174c5d
[INFO 2023-08-30T07:19:51.691899882+00:00 kernel.cc:1046] Use fast generic engine


i:72


[INFO 2023-08-30T07:19:52.96619698+00:00 kernel.cc:1214] Loading model from path /tmp/tmpt49qnk87/model/ with prefix b0abc6021d4d458f
[INFO 2023-08-30T07:19:52.980455594+00:00 kernel.cc:1046] Use fast generic engine


i:73


[INFO 2023-08-30T07:19:54.044835284+00:00 kernel.cc:1214] Loading model from path /tmp/tmpd7zb1l88/model/ with prefix 1d2a768d144d47de
[INFO 2023-08-30T07:19:54.051368807+00:00 kernel.cc:1046] Use fast generic engine


i:74


[INFO 2023-08-30T07:19:55.255810278+00:00 kernel.cc:1214] Loading model from path /tmp/tmptg1p4gt1/model/ with prefix a62d41ca838c4714
[INFO 2023-08-30T07:19:55.266447961+00:00 kernel.cc:1046] Use fast generic engine


i:75


[INFO 2023-08-30T07:19:56.385918063+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdqu2jil3/model/ with prefix a94ef94175bb4b00
[INFO 2023-08-30T07:19:56.393404465+00:00 kernel.cc:1046] Use fast generic engine


i:76


[INFO 2023-08-30T07:19:57.398794932+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfpg520sf/model/ with prefix 154dc12ce5054244
[INFO 2023-08-30T07:19:57.402664405+00:00 kernel.cc:1046] Use fast generic engine


i:77


[INFO 2023-08-30T07:19:58.431755179+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0ozed3ui/model/ with prefix 4c3378fa08734dc9
[INFO 2023-08-30T07:19:58.436272092+00:00 kernel.cc:1046] Use fast generic engine


i:78


[INFO 2023-08-30T07:19:59.535881022+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4w4mwnet/model/ with prefix 0ea59835033e4034
[INFO 2023-08-30T07:19:59.542296027+00:00 kernel.cc:1046] Use fast generic engine


i:79


[INFO 2023-08-30T07:20:00.630253082+00:00 kernel.cc:1214] Loading model from path /tmp/tmpytk9q7te/model/ with prefix 9422e755c0f845c1
[INFO 2023-08-30T07:20:00.6365468+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-30T07:20:00.636594078+00:00 kernel.cc:1046] Use fast generic engine


i:80


[INFO 2023-08-30T07:20:01.769025767+00:00 kernel.cc:1214] Loading model from path /tmp/tmppm5gpijk/model/ with prefix 2915f4c1f24747e3
[INFO 2023-08-30T07:20:01.777899799+00:00 kernel.cc:1046] Use fast generic engine


i:81


[INFO 2023-08-30T07:20:03.651803565+00:00 kernel.cc:1214] Loading model from path /tmp/tmpoxno8cut/model/ with prefix 1fa0cd5b62fb43d1
[INFO 2023-08-30T07:20:03.66515133+00:00 kernel.cc:1046] Use fast generic engine


i:82


[INFO 2023-08-30T07:20:05.366190455+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwah3luat/model/ with prefix 1432a949d5d94ed3
[INFO 2023-08-30T07:20:05.376730299+00:00 kernel.cc:1046] Use fast generic engine


i:83


[INFO 2023-08-30T07:20:07.11180569+00:00 kernel.cc:1214] Loading model from path /tmp/tmpr1969utq/model/ with prefix fa16f7fe549744c5
[INFO 2023-08-30T07:20:07.120686021+00:00 kernel.cc:1046] Use fast generic engine


i:84


[INFO 2023-08-30T07:20:09.0307224+00:00 kernel.cc:1214] Loading model from path /tmp/tmpg5s6g0n0/model/ with prefix 5bcb5f39c75d42b1
[INFO 2023-08-30T07:20:09.045794172+00:00 kernel.cc:1046] Use fast generic engine


i:85


[INFO 2023-08-30T07:20:10.539094945+00:00 kernel.cc:1214] Loading model from path /tmp/tmps7w774li/model/ with prefix 03a8b9e55be64977
[INFO 2023-08-30T07:20:10.544995913+00:00 kernel.cc:1046] Use fast generic engine


i:86


[INFO 2023-08-30T07:20:11.823206012+00:00 kernel.cc:1214] Loading model from path /tmp/tmpykd02q4w/model/ with prefix c86f7b5c19ce4662
[INFO 2023-08-30T07:20:11.838861885+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-30T07:20:11.838915324+00:00 kernel.cc:1046] Use fast generic engine


i:87


[INFO 2023-08-30T07:20:13.62765052+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3o8j14db/model/ with prefix 155867d7737a4e50
[INFO 2023-08-30T07:20:13.644729446+00:00 kernel.cc:1046] Use fast generic engine


i:88


[INFO 2023-08-30T07:20:14.829038832+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmang19ea/model/ with prefix 77603d9bc6ba4e12
[INFO 2023-08-30T07:20:14.839600479+00:00 kernel.cc:1046] Use fast generic engine


i:89


[INFO 2023-08-30T07:20:15.89571035+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4kkrbz9z/model/ with prefix 56569a57d9874117
[INFO 2023-08-30T07:20:15.900196148+00:00 kernel.cc:1046] Use fast generic engine


i:90


[INFO 2023-08-30T07:20:17.058271861+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8llhhhq_/model/ with prefix 08018faaaef44edb
[INFO 2023-08-30T07:20:17.067960882+00:00 kernel.cc:1046] Use fast generic engine


i:91


[INFO 2023-08-30T07:20:18.10279553+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7gxtp22i/model/ with prefix cfab6886ee844ebf
[INFO 2023-08-30T07:20:18.109128045+00:00 kernel.cc:1046] Use fast generic engine


i:92


[INFO 2023-08-30T07:20:19.435182209+00:00 kernel.cc:1214] Loading model from path /tmp/tmp79x6ksqh/model/ with prefix a7ca1f1903344fd4
[INFO 2023-08-30T07:20:19.456105587+00:00 kernel.cc:1046] Use fast generic engine


i:93


[INFO 2023-08-30T07:20:20.57951338+00:00 kernel.cc:1214] Loading model from path /tmp/tmpofbw4s58/model/ with prefix ce1fc3798ffd491d
[INFO 2023-08-30T07:20:20.588923178+00:00 kernel.cc:1046] Use fast generic engine


i:94


[INFO 2023-08-30T07:20:21.61424762+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvwzfi4ks/model/ with prefix 812c2695fa4c48e2
[INFO 2023-08-30T07:20:21.62033943+00:00 kernel.cc:1046] Use fast generic engine


i:95


[INFO 2023-08-30T07:20:22.757940836+00:00 kernel.cc:1214] Loading model from path /tmp/tmplxhn94za/model/ with prefix b44a99380d504c67
[INFO 2023-08-30T07:20:22.765723622+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-30T07:20:22.765764746+00:00 kernel.cc:1046] Use fast generic engine


i:96


[INFO 2023-08-30T07:20:23.882896578+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5kr6s2lp/model/ with prefix bd9905ce01294c0c
[INFO 2023-08-30T07:20:23.89220873+00:00 kernel.cc:1046] Use fast generic engine


i:97


[INFO 2023-08-30T07:20:24.905254375+00:00 kernel.cc:1214] Loading model from path /tmp/tmpd0jtxppl/model/ with prefix 13e96929aaf7445b
[INFO 2023-08-30T07:20:24.91018027+00:00 kernel.cc:1046] Use fast generic engine


i:98


[INFO 2023-08-30T07:20:25.976479498+00:00 kernel.cc:1214] Loading model from path /tmp/tmp349yssxl/model/ with prefix d88c9f0facb54626
[INFO 2023-08-30T07:20:25.983713987+00:00 kernel.cc:1046] Use fast generic engine


i:99


[INFO 2023-08-30T07:20:27.214337407+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqoowcibl/model/ with prefix 4e97719614d44aae
[INFO 2023-08-30T07:20:27.227348256+00:00 kernel.cc:1046] Use fast generic engine


Submission exported to /kaggle/working/submission.csv


# What is next

If you want to learn more about TensorFlow Decision Forests and its advanced features, you can follow the official documentation [here](https://www.tensorflow.org/decision_forests) 